Spring base sw

sw801733 opened the pull request:
- findAll 기능
findAll 의 경우 Book 테이블의 전부를 가져오는 기능으로
SQL 문이 변하지 않는 정적이므로 Statement 로 SQL 문을 수행하도록 구현


- save 기능
책 등록 기능 추가를 위해 이번에는 POST 요청으로 얻은 인자를 이용해야 하므로 PreparedStatement 를 사용


- findByIsbn 기능
마찬가지로 isbn 인자를 SQL 문으로 사용하기 때문에 PreparedStatement 를 사용

---

일단 모든 Controller 메서드에 대해서는 테스트 성공

NotableDeveloper committed:
empty branch init

NotableDeveloper committed:
Create README.md

NotableDeveloper committed:
init SpringBase

NotableDeveloper committed:
style : 디렉터리 생성

- Controller 클래스가 담길 디렉터리 생성
- Entity 클래스가 담길 디렉터리 생성
- Repository 클래스가 담길 디렉터리 생성
- Service 클래스가 담길 디렉터리 생성

NotableDeveloper committed:
feat : Book 클래스 작성

- 서비스의 핵심 엔티티인 Book 클래스 작성
- Book의 분류를 나타내는 BookCategory 클래스 작성
- Book의 상태를 나타내는 BookStatus 클래스 작성

NotableDeveloper committed:
feat : BookRepository interface 작성

BookRepository의 인터페이스 작성

NotableDeveloper committed:
feat : MemoryBookRepository 구현

- BookRepository를 구현한 MemoryBookRepository를 작성
- Book 엔티티에 Lombok의 Getter, Builder 적용

NotableDeveloper committed:
feat : BookService 구현

MemoryBookRepository를 사용하는 BookService 구현

NotableDeveloper committed:
feat : BookController 구현

- BookService를 사용하는 BookController 구현
- Book 엔티티에 생성자 추가

NotableDeveloper committed:
도서 검색 긴ê¸ê°긴ê기능 구현

NotableDeveloper committed:
feat : RegisterBookDTO 추가

- BookController의 registryBook 메서드에 DTO로 입력 받도록 수정
- BookService도 DTO를 가공하여 Book을 만들도록 수정

NotableDeveloper committed:
feat : BaseResponse 구현

- 클라이언트에게 반환할 최상위 HTTP Response 클래스 작성
- 디렉터리 구조 일부 변경

NotableDeveloper committed:
feat : 예외 처리

- 예외를 처리하여 책 등록 및 조회 기능 모두 구현
- 추후 대여와 반납을 구현할 예정

NotableDeveloper committed:
feat : BorrowBookDTO 생성

- 책 대여를 위힌 요청 DTO인 BorrowBookDTO 생성
- BookStatus의 Borrawing을 Borrowing으로 수정
- BookController의 책 대여에 관한 메서드 최초 작성

NotableDeveloper committed:
feat : 책 대출 기능 구현
- 책을 대출하는 기능을 구현
- BookRepository에 최소한의 Method만 남기고 모두 삭제

NotableDeveloper committed:
feat : 책 반납 기능 구현

책을 반납하는 기능을 구현

NotableDeveloper committed:
feat : 책의 이름으로 검색하는 기능 추가

- 책의 이름을 검색하는 기능을 구현
- 전체적인 코드 스타일 적용

NotableDeveloper committed:
feat : StatementBookRepository 구현

- MariaDB와 연동을 하기 위해 의존성 추가
- MariaDB와 Service 클래스 간의 통신을 위한 StatementBookRepository 구현

NotableDeveloper committed:
refact : JdbcBookRepository로 수정
기존의 StatementBookRepository를 JdbcBookRepository로 이름 수정

NotableDeveloper committed:
Merge pull request #13 from ACS-Spring-Study/SpringBase-sj

JdbcRepository create

NotableDeveloper committed:
fix : MariaDB 연동

- JdbcBookRepository을 Service에 주입하도록 수정
- Quantifier를 Service에만 적용

sw801733 committed:
feat : DB 커넥션 연결 확인 출력문 추가

sw801733 committed:
feat : findAll 기능을 위한 Statement 추가

sw801733 committed:
feat : save 기능을 위한 PreparedStatement 추가

PreparedStatement 를 이용해서 insert 문을 통한 책 등록 기능 구현

sw801733 committed:
feat : findByIsbn 기능을 위한 PreparedStatement 추가

isbn 을 통한 책 조회를 위한 기능 추가

sw801733 committed:
feat : existsByIsbn 기능을 위한 PreparedStatement 추가

이미 있는 isbn 을 가진 책이 있는지 확인하는 기능을 추가

sw801733 committed:
refact : Response 를 위한 Book 을 set 하는 메서드 추가

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -72,23 +84,23 @@ public Book findByISBN(String isbn) {
+
+        Book book = setBook(resultSet);
 
         System.out.println(book.getTitle() + " 책 조회 성공");
```
https://hudi.blog/do-not-use-system-out-println-for-logging/ 

위 링크 참고 바랍니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -21,6 +21,7 @@ public JdbcBookRepository(){
     try {
       connection = DriverManager.getConnection(DATABASE_URL);
       statement = connection.createStatement();
+      System.out.println("DB Connection Succeed");
```
https://hudi.blog/do-not-use-system-out-println-for-logging/

위 링크 참고하여 관련된 내용들 일괄적으로 적절하게 수정 바랍니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -44,6 +48,29 @@ public boolean existsByIsbn(String isbn) {
+        books.add(book);
+      }
+
+      resultset.close();
```
close 메서드는 try-catch의 finally 문에서 호출하는 것이 좋아보입니다. (예외가 발생하는 경우에도 close 할 수 있어야 함.)

또, static이 붙은 경우에는 그 특성을 잘 고려하여 close를 호출 할 것인지를 잘 생각해야 합니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -33,7 +33,30 @@ public JdbcBookRepository(){
   @Override
   public Book save(Book book) {
-    return null;
+    String sql = "Insert into Book values(NULL, ?, ?, ?, ?, ?)";
```
insert into Book (Column1, Column2, ..., ColumnN) values (?, ?, ... ?) 의 형식으로 쿼리를 구성하는 것이 좋아보입니다.

또, ID 값의 경우 Book 테이블 생성 과정에서 ID 값에 대한 auto_increment을 적용하였기 때문에 Null 값으로 넣을 필요는 없습니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -93,7 +93,27 @@ public Book findByISBN(String isbn) {
+
+      ResultSet resultSet = preparedStatement.executeQuery();
+
+      if (resultSet.next()) {
```
앞서 작성한 쿼리를 개선하면서 이 부분도 보다 더 간단하게 개선할 수 있을 것으로 보입니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -21,6 +21,18 @@ public class JdbcBookRepository implements BookRepository {
+    book.setTitle(resultSet.getString("title"));
+    book.setIsbn(resultSet.getString("isbn"));
+    book.setAuthor(resultSet.getString("author"));
+    book.setCategory(BookCategory.valueOf(resultSet.getString("book_category")));
```
book의 setter 메서드에 값을 입력할 때 내부적으로 valueOf와 getter 호출이 일어나고 있습니다. 이는 앞에서 간단한 변수로 빼두어 사용하면 간결할 것 같습니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -54,13 +66,13 @@ public Book save(Book book) {
       System.out.println(e.getMessage());
-      return null;
     }
+    return null;
```
메서드가 null을 반환하는 경우는 없어야 합니다. 이 사항은 모든 메서드에 일괄 적용해주시기 바랍니다.

NotableDeveloper commented on src/main/java/com/example/demo/repository/JdbcBookRepository.java:
```diff
@@ -93,7 +93,27 @@ public Book findByISBN(String isbn) {
   @Override
   public boolean existsByIsbn(String isbn) {
-    return false;
+    String sql = "SELECT * from Book where isbn = ?";
```
이 쿼리는 아래와 같이 개선할 수 있습니다.

SELECT isbn EXISTS (SELECT isbn FROM book WHERE isbn = ?)

위와 같이 exists와 서브쿼리를 사용하면 입력으로 주어진 isbn과 일치하는 레코드를 찾은 시점에 검색을 종료합니다.

그에 반하여 지금 작성한 쿼리는 입력으로 주어진 isbn과 일치하는 레코드를 찾아도 테이블의 모든 내용을 찾아가며 추가적으로 일치하는 레코드를 찾게 됩니다.

또, 지금처럼 단순 조회(= 값이 있는 지 혹은 없는 지와 같은 조회)의 경우, select 쿼리가 반환하려는 컬럼을 아래와 같은 와일드 카드(*)로 주게 되면, 가져오는 컬럼이 많아지게 되므로 비효율적입니다.

SELECT * from Book where isbn = ?

따라서, 아래처럼 하나의 컬럼만을 가져오는 것이 더 효율적일 것입니다.

SELECT isbn from Book where isbn = ?

sw801733 committed:
fix : sout 대신 logback 을 통한 로그 사용으로 변경

sw801733 committed:
refact : finally 에서 close 수행하도록 수정

sw801733 committed:
refact: save 메서드의 insert SQL 문 수정

Auto_Increment 로 id 값을 추가해 줄 필요가 없다.

sw801733 committed:
Merge remote-tracking branch 'origin/SpringBase-sw' into SpringBase-sw

# Conflicts:
#	src/main/java/com/example/demo/repository/JdbcBookRepository.java

sw801733 committed:
refact : setBook 의 메서드 인자 안 단순화

sw801733 committed:
refact : return null 대신 예외 발생

sw801733 committed:
refact : existsByIsbn 메서드의 SQL 문 성능 개선

sw801733 committed:
chore : JPA 의존성 추가

sw801733 committed:
feat : JPA 적용을 위한 JpaBookRepository 생성 및 적용

- JPA 적용을 위한 JpaBookRepository 생성
- Service 에 JpaBookRepository 적용
- JdbcRepository 의 MariaDB connect 부분 주석 처리

sw801733 committed:
chore : JPA 와  MariaDB 데이터베이스 연동을 위한 환경 변수 설정

sw801733 committed:
feat : Book 클래스에 엔티티 매핑을 위한 어노테이션 추가

sw801733 committed:
feat : JpaBookRepository isbn 검색을 위한 메서드 추가

sw801733 committed:
fix : Book 클래스의 isbn 의 이름에 맞게 메서드 명 수정

sw801733 committed:
feat : JpaBookRepository title 검색을 위한 메서드 추가

sw801733 committed:
fix : JPA 구현에 맞게 findAllContainsTitle 로직 변경

sw801733 committed:
feat : JpaBookRepository author 검색을 위한 메서드 추가

sw801733 committed:
fix : JPA 구현에 맞게 findAllContainsAuthor 로직 변경

sw801733 committed:
fix : Book 엔티티 ID 에 Auto increment 적용

sw801733 committed:
fix : JPA 요청을 통한 테이블 생성 옵션 none 으로 설정

sw801733 committed:
fix : ENUM 타입 저장을 위해 @Enumerated 추가

sw801733 committed:
feat : JpaBookRepository category 검색을 위한 메서드 추가

sw801733 committed:
fix : JPA 구현에 맞게 findAllByCategory 로직 변경

sw801733 committed:
fix : 책 대여와 반납의 상태 변화에 맞게 DB 업데이트 수행

반납에 경우 @Transactional을 대여에 경우 save 를 사용
둘다 업데이트 수행이 가능
